### PR TITLE
Add version on trie nodes

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- In order to have an endpoint that migrates trie nodes from a version to another, we need a way to know which nodes are saved as which version. 
  
## Proposed changes
- Add a new field to each type of trie node. A leaf node will remember it's own version, while branch nodes and extension nodes will save the version of their children

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: state/trackableDataTrie.go [hunk 1]
```diff
@@ -193,34 +204,46 @@ func (tdaw *trackableDataTrie) updateTrieWithAutoBalancing() (map[string][]byte,
...
-	return []byte(key), oldVal, nil
+	return common.TrieData{
+		Key:     []byte(key),
+		Value:   oldVal,
+		Version: common.NotSpecified,
+	}, nil
 }
 
-func (tdaw *trackableDataTrie) updateValInTrie(key []byte, val []byte) error {
+func (tdaw *trackableDataTrie) updateValInTrie(key []byte, val []byte, selfDataTrie dataTrie) error {
```
Comment (ssd04) [inline state/trackableDataTrie.go:244]: i'm thinking if we should consider renaming this `updateValInTrieWithAutobalancing` or something,  to differentiate from v1

File: testscommon/trie/trieStub.go [hunk 2]
```diff
@@ -86,6 +87,15 @@ func (ts *TrieStub) Update(key, value []byte) error {
 	return errNotImplemented
 }
 
+// UpdateWithVersion -
+func (ts *TrieStub) UpdateWithVersion(key []byte, value []byte, version common.TrieNodeVersion) error {
+	if ts.UpdateCalled != nil {
+		return ts.UpdateCalled(key, value)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply testscommon/trie/trieStub.go:93]: Changed.

File: state/trackableDataTrie.go [hunk 3]
```diff
@@ -172,18 +183,18 @@ func (tdaw *trackableDataTrie) updateTrieV1() (map[string][]byte, error) {
 	return oldValues, nil
 }
 
-func (tdaw *trackableDataTrie) updateTrieWithAutoBalancing() (map[string][]byte, error) {
-	oldValues := make(map[string][]byte)
+func (tdaw *trackableDataTrie) updateTrieWithAutoBalancing(dtr dataTrie) ([]common.TrieData, error) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go:186]: Added.

File: trie/branchNode.go [hunk 4]
```diff
@@ -945,6 +973,36 @@ func (bn *branchNode) collectStats(ts common.TrieStatisticsHandler, depthLevel i
 	return nil
 }
 
+func (bn *branchNode) getVersion() (common.TrieNodeVersion, error) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/branchNode.go:976]: Added.

**Review by ssd04 — APPROVED** [review]

File: state/trackableDataTrie.go [hunk 5]
```diff
@@ -137,23 +138,33 @@ func (tdaw *trackableDataTrie) SaveDirtyData(mainTrie common.Trie) (map[string][
...
 	}
 
-	return tdaw.updateTrieV1()
+	return tdaw.updateTrieV1(dtr)
 }
 
-func (tdaw *trackableDataTrie) updateTrieV1() (map[string][]byte, error) {
-	oldValues := make(map[string][]byte)
+func (tdaw *trackableDataTrie) updateTrieV1(selfDataTrie dataTrie) ([]common.TrieData, error) {
+	oldValues := make([]common.TrieData, 0)
```
Comment (AdoAdoAdo) [inline state/trackableDataTrie.go:154]: also here?
`make([]common.TrieData, len(tdaw.dirtyData))`

Comment (BeniaminDrasovean) [reply state/trackableDataTrie.go]: Done.

**Review by AdoAdoAdo — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]

**Pull request merged**